File: src/main/java/me/lokka30/levelledmobs/listeners/ChunkLoadListener.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.listeners;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.util.Utils;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.misc.QueueItem;
import me.lokka30.levelledmobs.wrappers.SchedulerWrapper;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.world.ChunkLoadEvent;

/**
 * Listens for when chunks are loaded and processes any mobs accordingly Needed for server startup
 * and for mostly passive mobs when players are moving around
 *
 * @author stumper66
 * @since 2.4.0
 */
public class ChunkLoadListener implements Listener {

    private final LevelledMobs main;

    public ChunkLoadListener(final LevelledMobs main) {
        this.main = main;
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
    public void onChunkLoad(final ChunkLoadEvent event) {

        if (!main.helperSettings.getBoolean(main.settingsCfg,
            "ensure-mobs-are-levelled-on-chunk-load", true)) {
            return;
        }

        // Check each entity in the chunk
        for (final Entity entity : event.getChunk().getEntities()) {

            // Must be a *living* entity
            if (!(entity instanceof final LivingEntity livingEntity)) {
                continue;
            }

            checkEntity(livingEntity, event);
        }
    }

    private void checkEntity(final LivingEntity livingEntity, final ChunkLoadEvent event){
        final SchedulerWrapper wrapper = new SchedulerWrapper(livingEntity, () -> {
            final LivingEntityWrapper lmEntity = LivingEntityWrapper.getInstance(
                    livingEntity, main);

            if (main.levelManager.doCheckMobHash && Utils.checkIfMobHashChanged(lmEntity)) {
                lmEntity.reEvaluateLevel = true;
                lmEntity.isRulesForceAll = true;
                lmEntity.wasPreviouslyLevelled = lmEntity.isLevelled();
            }
            else if (lmEntity.isLevelled()) {
                lmEntity.free();
                return;
            }

            main.mobsQueueManager.addToQueue(new QueueItem(lmEntity, event));
            lmEntity.free();
        });

        wrapper.runDirectlyInBukkit = true;
        wrapper.run();
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/listeners/ChunkLoadListener.kt
package io.github.arcaneplugins.levelledmobs.listeners

import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.misc.QueueItem
import io.github.arcaneplugins.levelledmobs.util.Utils
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import io.github.arcaneplugins.levelledmobs.wrappers.SchedulerWrapper
import org.bukkit.entity.LivingEntity
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.world.ChunkLoadEvent

/**
 * Listens for when chunks are loaded and processes any mobs accordingly Needed for server startup
 * and for mostly passive mobs when players are moving around
 *
 * @author stumper66
 * @since 2.4.0
 */
class ChunkLoadListener : Listener {
    private var ensureMobsAreLevelledOnChunkLoad = true

    fun load(){
        ensureMobsAreLevelledOnChunkLoad = LevelledMobs.instance.helperSettings.getBoolean(
            "ensure-mobs-are-levelled-on-chunk-load", true
        )
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
    fun onChunkLoad(event: ChunkLoadEvent) {
        if (!ensureMobsAreLevelledOnChunkLoad) return

        // Check each entity in the chunk
        for (entity in event.chunk.entities) {
            // Must be a *living* entity

            if (entity !is LivingEntity) {
                continue
            }

            checkEntity(entity, event)
        }
    }

    private fun checkEntity(livingEntity: LivingEntity, event: ChunkLoadEvent) {
        val lmEntity = LivingEntityWrapper.getInstance(livingEntity)
        livingEntity.scheduler.run(LevelledMobs.instance, { task ->
            val wrapper = SchedulerWrapper(livingEntity) {
                if (LevelledMobs.instance.levelManager.doCheckMobHash && Utils.checkIfMobHashChanged(lmEntity)) {
                    lmEntity.reEvaluateLevel = true
                    lmEntity.isRulesForceAll = true
                    lmEntity.wasPreviouslyLevelled = lmEntity.isLevelled
                } else if (lmEntity.isLevelled) {
                    lmEntity.free()
                    return@SchedulerWrapper
                }

                LevelledMobs.instance.mobsQueueManager.addToQueue(QueueItem(lmEntity, event))
                lmEntity.free()
            }

            lmEntity.buildCacheIfNeeded()
            lmEntity.inUseCount.getAndIncrement()
            wrapper.runDirectlyInBukkit = true
            wrapper.run()
        }, null)
    }
}